cpp/src/parquet/arrow/reader.h

```diff
@@ -175,6 +177,22 @@ class PARQUET_EXPORT FileReader {
       const std::vector<int>& row_group_indices, const std::vector<int>& column_indices,
       std::unique_ptr<::arrow::RecordBatchReader>* out) = 0;
 
+  /// \brief Return a generator of record batch vectors, where each vector represents
+  ///     the contents of a row group from row_group_indices, whose columns are selected
+  ///     by column_indices.
+  ///
+  /// An empty optional indicates the end of the generator.
+  ///
+  /// Note that the ordering in row_group_indices and column_indices matter. FileReaders
+  /// must outlive their generators.
+  ///
+  /// \returns error Result if either row_group_indices or column_indices contains an
+  ///     invalid index
+  virtual ::arrow::Result<
```

Comment:
Answering your question I agree it would be better to have `AsyncGenerator<RecordBatch>` for consistency with the other readers.  You can use `MakeVectorGenerator` to get `AsyncGenerator<AsyncGenerator<RecordBatch>>`.  Then apply `MakeConcatMapGenerator` to the result to get to `AsyncGenerator<RecordBatch>`